code/modules/defenses/sentry_computer.dm

```diff
@@ -0,0 +1,439 @@
+	.["camera_target"] = null
+
+	var/index = 1
+
+	for(var/sentry in paired_sentry)
+		var/list/sentry_holder = list()
+		var/obj/structure/machinery/defenses/defence = sentry
+		if(current == defence)
+			.["camera_target"] = index
+		index++
+
+		sentry_holder["area"] = get_area(defence)
+		sentry_holder["active"] = defence.turned_on
+		sentry_holder["nickname"] = defence.nickname
+		sentry_holder["camera_available"] = defence.has_camera && defence.placed
+		sentry_holder["selection_state"] = list()
+		sentry_holder["kills"] = defence.kills
+		sentry_holder["iff_status"] = defence.faction_group
+		sentry_holder["health"] = defence.health
+		for(var/category in defence.selected_categories)
+			sentry_holder["selection_state"] += list(list("[category]", defence.selected_categories[category]))
+		if(istype(defence, /obj/structure/machinery/defenses/sentry))
+			var/obj/structure/machinery/defenses/sentry/sentrygun = sentry
+			sentry_holder["rounds"] = sentrygun.ammo.current_rounds
+			sentry_holder["max_rounds"] = sentrygun.ammo.max_rounds
+			sentry_holder["engaged"] = length(sentrygun.targets)
+
+		.["sentry"] += list(sentry_holder)
+
+/obj/item/device/sentry_computer/tgui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	update_active_camera()
+	if (!ui)
+		// Register map objects
+		user.client.register_map_obj(cam_background)
+		user.client.register_map_obj(cam_screen)
+		ui = new(user, src, "SentryGunUI", name)
+		ui.open()
+
+/obj/item/device/sentry_computer/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
+	. = ..()
+	if(.)
+		return
+	if(!skillcheck(usr, SKILL_ENGINEER, SKILL_ENGINEER_ENGI))
+		to_chat(usr, SPAN_WARNING("You are not authorised to configure the sentry."))
+		return
+	if(params["index"])
+		// the action represents a sentry
+		var/sentry_index = params["index"]
+		if(paired_sentry[sentry_index])
+			var/result = paired_sentry[sentry_index].update_choice(usr, action, params["selection"])
+			if(result)
+				playsound(src, get_sfx("terminal_button"), 25, FALSE)
+				return TRUE
+			switch(action)
+				if("set-camera")
+					current = paired_sentry[sentry_index]
+					playsound(src, get_sfx("terminal_button"), 25, FALSE)
+					update_active_camera()
+					return TRUE
+				if("ping")
+					paired_sentry[sentry_index].identify()
+					return FALSE
+
+	switch(action)
+		if("screen-state")
+			screen_state = params["state"]
+			return FALSE
+		if("clear-camera")
+			current = null
+			playsound(src, get_sfx("terminal_button"), 25, FALSE)
+			update_active_camera()
+			return TRUE
+		if("ui-interact")
+			playsound(src, get_sfx("terminal_button"), 25, FALSE)
+			return FALSE
+
+
+/obj/item/device/sentry_computer/proc/show_camera_static()
+	cam_screen.vis_contents.Cut()
+	last_camera_turf = null
+	cam_background.icon_state = "scanline2"
+	cam_background.fill_rect(1, 1, DEFAULT_MAP_SIZE, DEFAULT_MAP_SIZE)
+
+
+/obj/item/device/sentry_computer/proc/update_active_camera()
```

Comment:
is there a genericized way to do this that's already done via overwatch or similar?

Reply:
The camera code is scattered throughout the code, there isn't a centralised way to do it. I was thinking of refactoring this all into a datum but just wanting to get this merged at this point